sa/sa_test.go

```diff
@@ -2574,3 +2578,139 @@ func TestIncidentsForSerial(t *testing.T) {
+	s.sfiChan <- sfiData{serial, nil}
+	return nil
+}
+
+// Context returns the context originally passed by the client.
+func (s sfiServer) Context() context.Context {
+	return s.context
+}
+
+func (s sfiServer) sendErrAndClose(err error) error {
+	s.sfiChan <- sfiData{nil, err}
+	close(s.sfiChan)
+	return nil
+}
+
+func makeinmemSAGRPCStreamAdapter(ctx context.Context) (sfiClient, sfiServer) {
+	sfiDataChan := make(chan sfiData)
+	return sfiClient{sfiChan: sfiDataChan}, sfiServer{sfiChan: sfiDataChan, context: ctx}
+}
+
+func (s inmemSA) SerialsForIncident(ctx context.Context, req *sapb.SerialsForIncidentRequest, _ ...grpc.CallOption) (sapb.StorageAuthority_SerialsForIncidentClient, error) {
+	client, server := makeinmemSAGRPCStreamAdapter(ctx)
+	go func() {
+		err := s.Impl.SerialsForIncident(req, server)
+		if err != nil {
+			server.sendErrAndClose(err)
+		} else {
+			server.sendErrAndClose(io.EOF)
+		}
+	}()
+	return client, nil
+}
+
+func TestSerialsForIncident(t *testing.T) {
+	sa, _, cleanUp := initSA(t)
+	defer cleanUp()
+
+	// Attempt to query incident serials which don't exist.
+	isa := inmemSA{Impl: sa}
+	client, err := isa.SerialsForIncident(context.Background(), &sapb.SerialsForIncidentRequest{
+		IncidentTable: "incident_foo"})
+	test.AssertNotError(t, err, "Error calling SerialsForIncident")
+
+	// Should only get `io.EOF` error as there are no rows in the requested
+	// incident table.
+	for {
+		_, err := client.Recv()
+		test.AssertError(t, err, "Expected io.EOF error")
+		test.AssertErrorIs(t, err, io.EOF)
+		break
+	}
+
+	// Add a incident serial rows to the incident table.
+	for _, i := range []string{"1335", "1336", "1337", "1338"} {
+		mrand.Seed(time.Now().Unix())
+		randInt := func() int64 { return mrand.Int63() }
+		_, err := sa.dbMap.Exec(
+			fmt.Sprintf("INSERT INTO incident_foo (%s) VALUES ('%s', %d, %d, '%s')",
+				"serial, registrationID, orderID, lastNoticeSent",
+				i,
+				randInt(),
+				randInt(),
+				sa.clk.Now().Add(time.Hour*24*7).Format("2006-01-02 15:04:05"),
+			),
+		)
+		test.AssertNotError(t, err, fmt.Sprintf("Error while inserting row for '%s' into incident table", i))
+	}
+
+	// Attempt to query for the rows we just added.
+	client, err = isa.SerialsForIncident(
+		context.Background(), &sapb.SerialsForIncidentRequest{
+			IncidentTable: "incident_foo"})
+	test.AssertNotError(t, err, "Error getting serials for incident")
+
+	// Ensure that we receive all four of the expected serials.
+	var received int
+	for {
+		serial, err := client.Recv()
+		if err != nil {
+			// An `io.EOF` error is expected when we've received all four
+			// serials.
+			if errors.Is(err, io.EOF) {
+				// Ensure that we received all four.
+				test.AssertEquals(t, received, 4)
+				break
+			}
+			test.AssertNotError(t, err, "Received error other than io.EOF")
+		}
+		switch serial.Serial {
+		case "1335":
+			received++
+		case "1336":
+			received++
+		case "1337":
+			received++
+		case "1338":
+			received++
```

Comment:
This has a surprising property that we could receive the serial "1335" over and over again, and not notice. If we're going to the trouble of checking for specific serial numbers, it would be better to build a map of them, and check-and-delete as we go; if ever a key is absent, we fail.